File: LeetCode/39. Combination Sum.cpp
// Author : @AdarshGaur

class Solution {
public:
    vector<vector<int>> ans;
    
    void solve(vector<int>& C, vector<int>& store, int i, int need){
        if(i == C.size() or need < 0) return;
        if(need == 0){
            ans.push_back(store);
            return;
        }
        
        store.push_back(C[i]);
        solve(C, store, i, need - C[i]);
        store.pop_back();
        
        solve(C, store, i+1, need);
    }
    
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> store;
        solve(candidates, store, 0, target);
        return ans;
    }
};

File: LeetCode/39. Combination Sum.py
#Author: @AdarshGaur

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        self.backtrack(candidates, ans, target, 0)
        return ans
    
    def backtrack(self, A, ans, need, indx, path=[]):
        if need == 0:
            ans.append(path)
            return
        elif need < 0 or indx == len(A):
            return 
        
        for i in range(indx, len(A)):
            self.backtrack(A, ans, need - A[i], i, path + [A[i]])
        return
